### PR TITLE
[5.7] Multiple mail connections

### PR DESCRIPTION
**Pull request by Propaganistas**
I'm currently working on a project where multiple mails should be sent in a single request, each requiring their own SMTP settings and there's even one that could be sent using another driver. At the moment Laravel constructs the Mailer during first initialisation and then doesn't allow switching driver or even configuration like SMTP settings that easily.

I think the Mail component could benefit greatly from adopting a similar setup as the Database component: 1) multiple connections can be defined, which accommodates the need for multiple configurations using the same driver but with different settings; and 2) the "connection" can be chosen on runtime without any hassle.

For example, [this could be the `config/mail.php` file](https://github.com/Propaganistas/laravel/blob/1950c867ba638017741c27c7208c781dab2522d6/config/mail.php#L33), and then mails could be sent as follows:
```php
Mail::send(...); // Uses the default connection.
Mail::connection('other_smtp')->send(...); // Uses the "other_smtp" connection
Mail::connection('sparkpost')->send(...); // Uses the "sparkpost" connection
```

I'm eager to learn your insights.

### CODE REVIEW

File: src/Illuminate/Mail/MailManager.php
Comment (GrahamCampbell) [inline src/Illuminate/Mail/MailManager.php]: string|null

**Comment by GrahamCampbell** [comment]
The manager is missing the ability to extend with more drivers. Might be cool to go as far as having a custom driver like the LogManager does.

**Comment by Propaganistas** [comment]
Yeah as I mentioned in the disclaimer it's far from perfect (and perhaps isn't even working in its current state). Just wanted to reach out first to avoid spending too much time on a concept that gets rejected.

**Comment by Propaganistas** [comment]
As an additional note: the DatabaseManager deviates from `Illuminate\Support\Manager` in the sense that a driver is defined on connection level, while the `Manager` directly asks for and returns various drivers.

If the overall concept is ok, perhaps this is a good moment to introduce a second Manager-like Support class (e.g. `ConnectionManager`) and let `DatabaseManager` and `MailManager` extend from it?

**Comment by decadence** [comment]
It's certainly the thing Laravel is missing.
When I need this I use [this workaround](https://stackoverflow.com/questions/26546824/multiple-mail-configurations/26547428)

**Comment by taylorotwell** [comment]
I'm generally fine with the idea behind it if you want to polish it up and test it.

**Comment by regiszanandrea** [comment]
@decadence Me too, I have to override the method `send`of Mailable Class.

**Comment by Propaganistas** [comment]
Awaiting feedback on the referenced PR that adds the mentioned `ConnectionManager` support class.

**Comment by taylorotwell** [comment]
I don't want to get bogged down with new abstractions and generalizations yet. Let's just focus on the mail aspect.

**Comment by Propaganistas** [comment]
Here's a first take on this. I'm unsure if the tests are fine and could use some advice on this.
[Here's the PR](https://github.com/laravel/laravel/pull/4589) for changes in the default config file.

Looking forward to feedback.

File: src/Illuminate/Mail/MailManager.php
Comment (victorlap) [inline src/Illuminate/Mail/MailManager.php]: Should me mailer connection

File: src/Illuminate/Mail/MailManager.php
Comment (TBlindaruk) [inline src/Illuminate/Mail/MailManager.php]:
```PHP
public function connection(?string $name = null)
```

File: src/Illuminate/Mail/MailManager.php
Comment (TBlindaruk) [inline src/Illuminate/Mail/MailManager.php]:
```PHP
protected function makeConnection(string $name)
```

File: src/Illuminate/Mail/MailManager.php
Comment (TBlindaruk) [inline src/Illuminate/Mail/MailManager.php]:
```PHP
 protected function configuration(string $name)
```

File: src/Illuminate/Mail/MailManager.php
Comment (TBlindaruk) [inline src/Illuminate/Mail/MailManager.php]:
```PHP
protected function makeTransport(string $driver, array $config)
```

File: src/Illuminate/Mail/MailManager.php
Comment (TBlindaruk) [inline src/Illuminate/Mail/MailManager.php]:
```PHP
public function setDefaultConnection(string $name)
```

File: src/Illuminate/Mail/MailServiceProvider.php
Comment (TBlindaruk) [inline src/Illuminate/Mail/MailServiceProvider.php]: Please add phpdoc
```PHP 
@see MailManager
```

**Review by TBlindaruk — CHANGES_REQUESTED** [review]

File: src/Illuminate/Mail/MailManager.php
Comment (deleugpn) [reply src/Illuminate/Mail/MailManager.php]: The framework code doesn't make use of primitive type hints.

File: src/Illuminate/Mail/MailManager.php
Comment (GrahamCampbell) [reply src/Illuminate/Mail/MailManager.php]: Please don't add the string type hint.

File: src/Illuminate/Mail/MailManager.php
Comment (GrahamCampbell) [reply src/Illuminate/Mail/MailManager.php]: Please don't add the string type hint.

File: src/Illuminate/Mail/MailManager.php
Comment (GrahamCampbell) [reply src/Illuminate/Mail/MailManager.php]: No.

File: src/Illuminate/Mail/MailServiceProvider.php
Comment (Propaganistas) [reply src/Illuminate/Mail/MailServiceProvider.php]: The framework uses `@see` only at class level.

**Comment by brunogaspar** [comment]
Hey @Propaganistas 

This looks good overall, not sure if you plan on working on this a bit further so we can get this merged soon?

I've a question though, would this allow each connection to have it's send `from.name` and `from.address` instead of using the global one?
The idea would be to still have the global one though, but been able to override it per connection i guess.

**Comment by Propaganistas** [comment]
Hi @brunogaspar, I'm still waiting for some feedback as well. I think the core team wants to finish up some other stuff before merging this one.

Currently allowing each connection to specify its own `from` address isn't implemented but I really like the idea. Gonna take a shot at it. Thanks!

**Comment by brunogaspar** [comment]
@Propaganistas

> Currently allowing each connection to specify its own from address isn't implemented but I really like the idea. Gonna take a shot at it. Thanks!

That's cool :)

> I'm still waiting for some feedback as well.

I'm going to give this PR a try really soon and will leave some feedback after.

File: src/Illuminate/Mail/MailManager.php
Comment (goodevilgenius) [reply src/Illuminate/Mail/MailManager.php]: I've never been able to find Laravel's specific code style guidelines. Could you provide a link to where it prohibits scalar type hints?

Comment (devcircus) [reply src/Illuminate/Mail/MailManager.php]: There's nothing official that explicitly prohibits them, they're just not used elsewhere in the framework (except in the case of extending a third party dependency that uses them). Generally, Laravel follows psr2 with a few exceptions, like spaces in docblocks, as referenced [here](https://laravel.com/docs/5.6/contributions#coding-style).

Comment (goodevilgenius) [reply src/Illuminate/Mail/MailManager.php]: But PSR-2 doesn't prohibit type-hinting primitives, so if the guideline is PSR-2, there's no specific reason to disallow it. If it is specifically disallowed, it should be mentioned in the link you just posted. If it's not specifically disallowed, there's no reason not to include them (type safety, in general, is a good thing).

**Comment by taylorotwell** [comment]
I think I'm going to hold off on this for now. I will try my own take at it sometime.

**Comment by jozeflambrecht** [comment]
Is this still on the list?

**Comment by tjmartin69** [comment]
Is this something still on the roadmap?

**Comment by nguyentranchung** [comment]
i'm waiting feature 🤔

**Pull request closed without merging**
